src/classes/CW_AudienceCSVUploadControllerTest.cls

```diff
@@ -0,0 +1,397 @@
+* 0.2 15/04/2014 Klea Kolaric	Changes due to Audience Group Schedule
+*
+*
+*/
+@isTest
+private class CW_AudienceCSVUploadControllerTest 
+{
+	static testMethod void quickScheduleTest()
+	{
+		// test that closest schedule is picked up by default
+		
+		Account acc = TestDataUtil.createPersonAccount();
+		insert acc;
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		
+		// insert default schedule
+		CW_Schedule__c sch1 = TestDataUtil.createSchedule(camp);
+		insert sch1;
+		// insert later schedule
+		CW_Schedule__c sch2 = TestDataUtil.createSchedule(camp);
+		sch2.Start_Date__c = sch1.Start_Date__c.addDays(+2);
+		sch2.End_Date__c = sch1.End_Date__c.addDays(-2);
+		sch2.Evaluation_Start_Date__c = sch2.Start_Date__c;
+		sch2.Evaluation_End_Date__c = sch2.End_Date__c;
+		insert sch2;
+		
+		// insert group
+		CW_Audience_Group__c audGr1 = TestDataUtil.createAudienceGroup(camp.Id);
+		audGr1.Name = group1;
+		audGr1.Is_Control_Group__c = false;
+		insert audGr1;
+		
+		CW_Audience_Group_Schedule__c ags1 = [Select Id from CW_Audience_Group_Schedule__c where Audience_Group__c =: audGr1.id and Schedule__c =:sch1.Id];
+		
+		Test.startTest();
+		CW_Audience_Group_Member__c agm = new CW_Audience_Group_Member__c();
+		agm.Audience_Group_LU__c = audGr1.Id;
+		agm.Contact__c = [select id from Contact where AccountId = :acc.Id limit 1].Id;
+		agm.Audience_Group_Schedule__c = ags1.Id;
+		insert agm;
+		Test.stopTest();
+		
+		System.assertEquals(1,[select count() from CW_Audience_Group_Member__c where id = :agm.Id and Audience_Group_Schedule__c = :ags1.Id]);
+	}
+
+	static testMethod void parseGroupsAndMembers()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+		insert camp;
+		
+		CW_Schedule__c sch = TestDataUtil.createSchedule(camp);
+		insert sch;
+		
+		Test.setCurrentPage(Page.CW_AudienceCSVUpload);
+		Apexpages.currentPage().getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,camp.Id);
+		
+		Test.startTest();
+		CW_AudienceCSVUploadController uc = new CW_AudienceCSVUploadController();
+		uc.selectedSchId = sch.id;
+		uc.csvFile = Blob.valueOf(groupCSVString);
+		uc.filename='Somefile.csv';
+		uc.uploadCSV();
+		//uc.returnToAudiencePage();
+		Test.stopTest();
+		
+		// 6 new accounts were added
+		system.assertEquals(6,[select count() from Contact where Account_Name__c != :null]);
+		// 5 new groupss were added
+		system.assertEquals(5,[select count() from CW_Audience_Group__c where Campaign__c = :camp.Id]);
+		// 2 members where added to group1
+		system.assertEquals(2,[select count() from CW_Audience_Group_Member__c where Audience_Group_LU__r.Name = :group1]);
+		// 1 members where added to group2
+		system.assertEquals(1,[select count() from CW_Audience_Group_Member__c where Audience_Group_LU__r.Name = :group2]);
+		// 1 members where added to group3
+		system.assertEquals(1,[select count() from CW_Audience_Group_Member__c where Audience_Group_LU__r.Name = :group3]);
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.